llama_index/schema.py

```diff
@@ -119,3 +147,303 @@ def extra_info_str(self) -> Optional[str]:
             return None
 
         return "\n".join([f"{k}: {str(v)}" for k, v in self.extra_info.items()])
+
+
+@dataclass
+class BaseIndexObject(DataClassJsonMixin):
+    """Base Index Object.
+
+    Generic abstract interface for retrievable objects
+
+    """
+
+    _id: str = ""
+    content: Any = None
+    relationships: Dict[DataRelationship, Any] = field(default_factory=dict)
+    embedding: Optional[List[float]] = None
+    object_hash: Optional[str] = None
+    weight: float = 1.0
+
+    """"
+    metadata fields
+    - injected as part of the text shown to LLMs as context
+    - used by vector DBs for metadata filtering
+
+    This must be a flat dictionary, 
+    and only uses str keys, and (str, int, float) values.
+    """
+    metadata: Dict[str, Any] = field(default_factory=dict)
+
+    # attributes are additional fields that do not influence query/retrieval
```

Comment:
Oh interesting, so if users want to specify metadata that's not passed into the text, they would define an attribute?

Reply:
I removed the attributes/metadata separation, and instead added a filter on the keys that are allowed to be used